// smooth-scroll/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = function (elem, target, options) {
  options = _extends({}, DEFAULTS, options);

  if (_interval) {
    clearInterval(_interval);
  }

  var startLocation = typeof elem.scrollTop !== 'undefined' ? elem.scrollTop : elem.pageYOffset;
  var distance = target - startLocation;
  var timeLapsed = 0;

  function stopAnimateScroll(position, flooredPosition) {
    if (position === target || flooredPosition === target || elem.innerHeight + flooredPosition >= getDocumentHeight()) {
      clearInterval(_interval);
      options.onScrolled();
    }
  }

  function loopAnimateScroll() {
    timeLapsed += 16;
    var percentage = timeLapsed / options.duration;
    percentage = percentage > 1 ? 1 : percentage;
    var position = startLocation + distance * options.easing(percentage);
    var flooredPosition = Math.floor(position);
    scroll(elem, flooredPosition);
    stopAnimateScroll(position, flooredPosition);
  }

  function startAnimateScroll() {
    _interval = setInterval(loopAnimateScroll, 16);
  }

  /**
   * Reset position to fix weird iOS bug
   * @link https://github.com/cferdinandi/smooth-scroll/issues/45
   */
  if (elem.scrollTop === 0 || elem.pageYOffset === 0) {
    scroll(elem, 0);
  }

  startAnimateScroll();
};

var DEFAULTS = {
  easing: easeInOutQuad,
  duration: 300,
  onScrolled: function onScrolled() {}
};

function easeInOutQuad(t) {
  return t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;
}

var _interval = void 0;

function getDocumentHeight() {
  return Math.max(document.body.scrollHeight, document.documentElement.scrollHeight, document.body.offsetHeight, document.documentElement.offsetHeight, document.body.clientHeight, document.documentElement.clientHeight);
}

function scroll(elem, value) {
  if (typeof elem.scrollTop !== 'undefined') {
    elem.scrollTop = value;
  } else {
    elem.scrollTo(0, value);
  }
}
